src/relentless/model/potential/bond.py

```diff
@@ -0,0 +1,425 @@
+import numpy
+
+from relentless import math
+
+from . import potential
+
+
+class BondParameters(potential.Parameters):
+    """Parameters for a bond potential."""
+
+    pass
+
+
+class BondPotential(potential.Potential):
+    r"""Abstract base class for a bond potential."""
+
+    pass
+
+
+class Harmonic(BondPotential):
+    r"""Harmonic bond potential.
+
+    .. math::
+
+        u(r) = \frac{k}{2} (r - r_0)^2
+
+    where :math:`r` is the distance between two bonded particles. The parameters
+    for each type are:
+
+    +-------------+--------------------------------------------------+-----------+
+    | Parameter   | Description                                      | Initial   |
+    +=============+==================================================+===========+
+    | ``k``       | potential constant                               |           |
+    +-------------+--------------------------------------------------+-----------+
+    | ``r0``      | rest length.                                     |           |
+    +-------------+--------------------------------------------------+-----------+
+
+    Parameters
+    ----------
+    type : str
+        Type.
+    name : str
+        Unique name of the potential. Defaults to ``__u[id]``, where ``id`` is the
+        unique integer ID of the potential.
+
+    Attributes
+    ----------
+    coeff : :class:`BondParameters`
+        Parameters of the potential for each type.
+
+    Examples
+    --------
+    Harmonic Bond::
+
+        >>> u = relentless.potential.bond.Harmonic("A")
+        >>> u.coeff["A"].update({'k': 1.0, 'r0': 1.0})
+
+    """
+
+    def __init__(self, types, name=None):
+        super().__init__(keys=types, params=("k", "r0"), name=name)
+
+    def energy(self, type, r):
+        """Evaluate bond energy."""
+        k = self.coeff[type]["k"]
+        r0 = self.coeff[type]["r0"]
+        return k / 2 * (r - r0) ** 2
+
+    def force(self, type, r):
+        """Evaluate bond force."""
+        k = self.coeff[type]["k"]
+        r0 = self.coeff[type]["r0"]
+        return -k * (r - r0)
+
+    def derivative(self, type, var, r):
+        r"""Evaluate bond derivative with respect to a variable."""
+        k = self.coeff[type]["k"]
+        r0 = self.coeff[type]["r0"]
+        if var == "k":
+            return (r - r0) ** 2 / 2
+        elif var == "r0":
+            return -k * (r - r0)
+        else:
+            raise ValueError("Unknown parameter")
+
+
+class FENE(BondPotential):
+    r"""The Finite Extensible Nonlinear Elastic (FENE) bond interaction.
+
+    .. math::
```

Comment:
Since we have realized for Tyla's project how important exclusions can be, I am wondering if we should enable the WCA part of the FENE interaction in case someone needs 1-2 exclusions for other parts of the model. We could do this in two ways:

* Add the WCA part here, with epsilon and sigma optional arguments defaulted to 0.
* Create a separate `FENEWCA` potential that derives from `FENE` and adds the `epsilon` and `sigma` parameters.

Reply:
I kind of like the WCA+FENE since we're going to be very explicit with exclusions and default arguments. I just found HOOMD's a bit hard my first time using it with their default exclusions.

Reply:
I agree, and it will be less code to only have one class. Our documentation is pretty explicit about defaults.